### PR TITLE
Implemented some viewport drawing functions

### PR DESCRIPTION
**Pull request by Gericom**
- sub_68B35F
- viewport_banner_paint_setup
-- This function is complete, but as long as 0x6C42D9 is not complete it
will cause glitches in the rendering of banners, because the wrong
(because localized) string length is used for calculating the scroll
position

### CODE REVIEW

File: src/interface/viewport.c [hunk 1]
```diff
@@ -844,6 +846,258 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+}
+
+//rct2: 0x0068B35F
+void sub_68B35F(int ax, int cx)
+{
+	if (ax < RCT2_GLOBAL(RCT2_ADDRESS_MAP_SIZE_UNITS, uint16_t) &&
+		cx < RCT2_GLOBAL(RCT2_ADDRESS_MAP_SIZE_UNITS, uint16_t) &&
+		ax >= 32 && cx >= 32)
+	{
+		RCT2_GLOBAL(0x141E9B4, uint32_t) = 0xFFFF;
```
Comment (adrian17) [inline src/interface/viewport.c]: It looks continuous, maybe use `memset`?

Comment (Gericom) [inline src/interface/viewport.c]: Yes. It's probably some kind of structure or something. I will have another look at it.

File: src/interface/viewport.c [hunk 2]
```diff
@@ -844,6 +846,258 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+			{
+				while (true)
+				{
+					element++;
+					if (bx < element->clearance_height) 
+						bx = element->clearance_height;
+					if (element->flags & MAP_ELEMENT_FLAG_LAST_TILE) break;
+				}
+			}
+			if ((element->type & 0x3C) == MAP_ELEMENT_TYPE_SURFACE &&
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Use the map element helper functions. There is one for getting the type and one for checking the last tile.

Comment (IntelOrca) [inline src/interface/viewport.c]: Rename bx to z or something and also do:
```z = max(z, element->clearance_height)```

**Comment by Gericom** [comment]
I added 0x6C42D9 now aswell, but for some reason the original viewport_banner_paint_setup is still called sometimes, causing localisation glitches. It would probably require some more implementation.

**Comment by duncanspumpkin** [comment]
signerror

Getting that along with the other glitches. I've been looking but I can't find anything that is calling the original banner code. Its a bit odd don't know why it jumps between the two at random.

**Comment by Gericom** [comment]
I thought that that glitch was fixed (your picture). I will have another look at that. I will make it call the original function and add a comment about that function that needs to be decompiled first.

**Comment by Gericom** [comment]
I fixed the glitch. Could you test it? If it's all okay, I will let it call the original function until  0x006E287B  is decompiled.

File: src/interface/viewport.c [hunk 3]
```diff
@@ -844,6 +846,342 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+	rct_drawpixelinfo* dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+	RCT2_GLOBAL(0x9DE570, uint8_t) = 0xC;
+	if (dpi->zoom_level > 1 || RCT2_GLOBAL(0x9DEA6F, uint8_t) & 1) return;
+	edx -= 16;
+	rct_scenery_entry* banner_scenery = g_bannerSceneryEntries[gBanners[map_element->properties.banner.index].type];
+	direction += map_element->properties.banner.position;
+	direction &= 3;
+	RCT2_GLOBAL(0x9DEA56, uint16_t) = edx + 2;
+	RCT2_GLOBAL(0x9DEA52, uint32_t) = RCT2_ADDRESS(0x98D884, uint32_t)[direction * 2];
+	int ebx = (direction << 1) + banner_scenery->image;
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: change ebx to image_id it will make it easier to follow.

File: src/interface/viewport.c [hunk 4]
```diff
@@ -844,6 +846,342 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+	ebx += (gBanners[map_element->properties.banner.index].colour << 19) | 0x20000000;
+	if (map_element->flags & 0x10)
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Add a comment saying if in ghost form. (i.e. when you are about to place)

**Comment by Gericom** [comment]
I will do that. But the glitch is solved, right?

**Comment by duncanspumpkin** [comment]
Yeah the glitch is solved.

File: src/interface/viewport.c [hunk 5]
```diff
@@ -844,6 +846,342 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+	while (true)
+	{
+		al = *format_result;
+		format_result++;
+		if (al == 0)
+		{
+			format_result = RCT2_ADDRESS(RCT2_ADDRESS_COMMON_STRING_FORMAT_BUFFER, char);
+			continue;
+		}
+		if (al < 0x9C && al >= 0x8E)
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: See https://github.com/IntelOrca/OpenRCT2/blob/develop/src/drawing/string.c#L1018-L1037
These offsets are known. (FORMAT_COLOUR_CODE_START, FORMAT_COLOUR_CODE_END)
I'm surprised that banners have the ability to change colours.

File: src/interface/viewport.c [hunk 6]
```diff
@@ -844,6 +846,342 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+		{
+			al -= 0x8E;
+			RCT2_GLOBAL(0x9D7A8C, uint8_t) = RCT2_ADDRESS(RCT2_GLOBAL(0x9FF048, uint32_t), uint8_t)[al * 4];
+			continue;
+		}
+		if (al < 0x20) continue;
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: If Control character. See above comment

File: src/interface/viewport.c (hunk 5 above)
Comment (Gericom) [inline src/interface/viewport.c]: I see.

You can set the text color in the sign dialog.

File: src/interface/viewport.c [hunk 7]
```diff
@@ -844,6 +846,342 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
...
+int sub_6C42D9(rct_string_id string_id, int scroll, int ebp)
+{
+	rct_drawpixelinfo* dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+	if (dpi->zoom_level != 0) return 0x626;
+	RCT2_GLOBAL(0x9D7A80, uint32_t)++;
+	uint32_t edx = 0xFFFFFFFF;
+	for (int i = 0; i < 0x20; i++)
+	{
+		uint8_t* unknown_pointer = RCT2_ADDRESS(0x9C3840, uint8_t) + 0xA12 * i;
+		if (edx >= *((int32_t*)(unknown_pointer + 0x0E)))
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Change this to uint32 since you have changed edx to unsigned.

Comment (Gericom) [inline src/interface/viewport.c]: I fixed that one already, but I did not yet push it.

File: src/interface/viewport.c (hunk 5 above)
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Oh yeah forgot about that.

**Comment by Gericom** [comment]
It should be ready to merge now.

**Pull request merged**